doc: Added serial recovery documentation (#1468)
mcu-tools/mcuboot: docs/readme-zephyr.md

Once MCUboot is built, this new keypair file (`mykey.pem` in this
example) can be used to sign images.

## Serial recovery

### Interface selection

A serial recovery protocol is available over either a serial port or an USB CDC ACM virtual serial port.
The SMP server implementation can be enabled by ``CONFIG_MCUBOOT_SERIAL=y``.
To set a type of an interface, use ``BOOT_SERIAL_DEVICE`` Kconfig choice, and select either ``CONFIG_BOOT_SERIAL_UART`` or ``CONFIG_BOOT_SERIAL_CDC_ACM`` value.

> Which interface belongs to the protocol shall be set by the devicetree-chosen node:

Author: In this sentence I'm explain how to configure the choice of the interface - not how to get the info.

- `zephyr,console` if serial port used.
- `zephyr,cdc-acm-uart` if virtual serial port used.

### Entering to the serial recovery mode

To enter the serial recovery mode, the device has to initiate rebooting, and a triggering event has to occur, for example, pressing a button.

> By default, the serial recovery GPIO pin active state desires entering the mode.

Reviewer: ```suggestion
By default, the serial recovery GPIO pin active state enters serial recovery mode.
```

The pin shall be assigned using the ``mcuboot_button0`` devicetree button alias.

Alternatively, MCUboot can wait for the prescribed duration to see if DFU is invoked by receiving an MCUmgr command.

> Select ``CONFIG_BOOT_SERIAL_WAIT_FOR_DFU=y`` to use this mode. ``CONFIG_BOOT_SERIAL_WAIT_FOR_DFU_TIMEOUT`` option defines

Author: Yes - this option determines the serial recovery mode used.

the amount of time the device will wait for the trigger.

### Direct image upload

> The SPM server implementation allows to select image number for DFU when invoking the MCUmgr `image upload` command.

Reviewer: ```suggestion
By default, the SMP server implementation will only use the the first slot, this can be changed to allow selecting the image number for DFU when invoking the MCUmgr `image upload` command when ``CONFIG_MCUBOOT_SERIAL_DIRECT_IMAGE_UPLOAD=y`` is enabled (note that multiple images can be supported by adjusting ``CONFIG_UPDATEABLE_IMAGE_NUMBER``).
```

The mapping of image number to partition is as follows:
* 0 and 1 - image-0, the primary slot of the first image.
* 2 - image-1, the secondary slot of the first image.
* 3 - image-2.

> * 4 - image-3.

Reviewer: Not sure this is right, as far as I know 0 is image 0 slot 0, 1 is image 0 slot 1, 2 is image 1 slot 0, 3 is image 1 slot 1, etc.
Can you double check this?

Reviewer: Ping on this comment

Author: This is input from @de-nordic. Mapping is not the same as for mcumgr-app. image-2 i image-3 can be reflected as given image slots...

> By default, the SMP server implementation will only use the the first slot, this can be changed to allow selecting the image number for DFU when invoking the MCUmgr `image upload` command when ``CONFIG_MCUBOOT_SERIAL_DIRECT_IMAGE_UPLOAD=y`` is enabled (note that multiple images can be supported by adjusting ``CONFIG_UPDATEABLE_IMAGE_NUMBER``).

Author: sentence have different meaning, see my correction

0 is a default upload target when no explicit selection is done.

> To enable this feature, use ``CONFIG_MCUBOOT_SERIAL_DIRECT_IMAGE_UPLOAD=y`` Kconfig option.

Reviewer: this line can go

### System-specific commands

Use ``CONFIG_ENABLE_MGMT_PERUSER=y`` Kconfig option to enable the following additional commands:

> * Storage erase - This command allows to erase the storage partition  (enable with ``CONFIG_BOOT_MGMT_CUSTOM_STORAGE_ERASE=y``).

Reviewer: ```suggestion
* Storage erase - This command allows erasing the storage partition (enable with ``CONFIG_BOOT_MGMT_CUSTOM_STORAGE_ERASE=y``).
```

> * Custom image list - This command allows to fetch versions and installation statuses (custom property) for all images (enable with ``CONFIG_BOOT_MGMT_CUSTOM_IMG_LIST=y``).

Reviewer: ```suggestion
* Custom image list - This command allows fetching version and installation status (custom properties) for all images (enable with ``CONFIG_BOOT_MGMT_CUSTOM_IMG_LIST=y``).
```

### In-place image decryption

Images uploaded by the serial recovery can be decrypted on the fly.
The encryption mechanism used in this case is ECIES using primitives described in the [ECIES encryption](encrypted_images.md#ecies-encryption) section. 

Enable support for this feature by using ``CONFIG_BOOT_SERIAL_ENCRYPT_EC256=y``.

### More configuration

> For details on other available configuration options for the serial recovery protocol, reach out to the Kconfig help.

Reviewer: ```suggestion
For details on other available configuration options for the serial recovery protocol, check the Kconfig options (e.g. by using ``menuconfig``).
```

Once this is done, this new keypair file (`mykey.pem` in this
example) can be used to sign images.

## Serial recovery

### interface selection

> Serial recovery protocol can be availeble over either the serial port or the USB CDC ACM virtual serial port.

Reviewer: ```suggestion
Serial recovery protocol can be available over either the serial port or the USB CDC ACM virtual serial port.
`

The SMP server implementation can be enabled by ``CONFIG_MCUBOOT_SERIAL=y``.

> Kind of interface used can be set using ``BOOT_SERIAL_DEVICE`` Kconfig chose: ``CONFIG_BOOT_SERIAL_UART`` or ``CONFIG_BOOT_SERIAL_CDC_ACM`` can be selected.

Reviewer: ```suggestion
The interface can be selected by using ``BOOT_SERIAL_DEVICE`` Kconfig choice: ``CONFIG_BOOT_SERIAL_UART`` or ``CONFIG_BOOT_SERIAL_CDC_ACM`` can be selected.
```

Which interface belongs to the protocol shall be set by the devicetree-chosen-node:
- `zephyr,console` if serial port used.
- `zephyr,cdc-acm-uart` if virtual serial port used.

### Entering to the serial recovery mode

> Mode can be entered once the device gets reboots and triggering event happens.

Reviewer: ```suggestion
Serial recovery mode can be entered once the device is rebooted or powered up and event is triggered.
```

By default the serial recovery GPIO pin active state desires entering the mode.
The pin shall be assigned using the ``mcuboot_button0`` devicetree button alias.

> Alternatively, the MCUboot can wait for a prescribed duration to see if DFU is invoked by receiving a mcumgr command.

Reviewer: ```suggestion
Alternatively, MCUboot can wait for a prescribed duration to see if DFU is invoked by receiving an MCUmgr command.
```

Select ``CONFIG_BOOT_SERIAL_WAIT_FOR_DFU=y`` to use this mode. ``CONFIG_BOOT_SERIAL_WAIT_FOR_DFU_TIMEOUT`` option defines
the amount of time the device will wait for the trigger.

### Direct image upload

> The implementation allows to select image number for DFU when invoke the mcumgr `image upload` command.

Reviewer: ```suggestion
The implementation allows to select image number for DFU when invoking the MCUmgr `image upload` command.
```

The mapping of image number to partition is as follows:
  0 -> default behavior, same as 1;
  1 -> image-0 (primary slot of the first image);
  2 -> image-1 (secondary slot of the first image);
  3 -> image-2;
  4 -> image-3.
Note that 0 is default upload target when no explicit selection is done.
This feature can be enabled using ``CONFIG_MCUBOOT_SERIAL_DIRECT_IMAGE_UPLOAD=y`` Kconfig option.

### System specific commands

Additional commands can be enabled using ``CONFIG_ENABLE_MGMT_PERUSER=y``.

> Following commands are added:
> - storage erase command, this command allows to erase the storage partition.
> - custom image list command, this command allow to fetch versions and installation statuses (custom property) for all images.

Reviewer: ```suggestion
Following commands can be enabled:
- storage erase command, this command allows erasing the storage partition (enable with ``CONFIG_BOOT_MGMT_CUSTOM_STORAGE_ERASE=y``).
- custom image list command, this command allow fetching versions and installation statuses (custom property) for all images (enable with ``CONFIG_BOOT_MGMT_CUSTOM_IMG_LIST=y``).
```

## Serial recovery

### Interface selection

A serial recovery protocol is available over either a serial port or an USB CDC ACM virtual serial port.
The SMP server implementation can be enabled by ``CONFIG_MCUBOOT_SERIAL=y``.

> To set a type of an interface, use ``BOOT_SERIAL_DEVICE`` Kconfig option and choose either ``CONFIG_BOOT_SERIAL_UART`` or ``CONFIG_BOOT_SERIAL_CDC_ACM``.

Author: ``BOOT_SERIAL_DEVICE`` is Kconfig choice.
sentence should stand:
>To set a type of an interface, use ``BOOT_SERIAL_DEVICE`` Kconfig choice, select either ``CONFIG_BOOT_SERIAL_UART`` or ``CONFIG_BOOT_SERIAL_CDC_ACM``.

Reviewer 2: ```suggestion
To set a type of an interface, use ``BOOT_SERIAL_DEVICE`` Kconfig choice, and select either ``CONFIG_BOOT_SERIAL_UART`` or ``CONFIG_BOOT_SERIAL_CDC_ACM`` value.
```

Which interface belongs to the protocol shall be set by the devicetree-chosen node:
- `zephyr,console` if serial port used.
- `zephyr,cdc-acm-uart` if virtual serial port used.

### Entering to the serial recovery mode

> The serial recovery mode can be entered once the device is rebooted or powered up and event is triggered

Author: The serial recovery mode can be entered once the device is rebooted and a triggering event happens.

Reviewer 2: ```suggestion
To enter the serial recovery mode, the device has to initiate rebooting, and a triggering event has to occur, for example, pressing a button.
```
As discussed on the chat.

Reviewer 2: Corrected as discussed on the chat.

and marked it as test/confirmed but MCUboot will not swap the images and
erasing the secondary slot from the zephyr application returns an error
because the slot is marked for upgrade.

## Serial recovery

### Interface selection

> A serial recovery protocol is available over either a hardware serial port or an USB CDC ACM virtual serial port.

Reviewer: ```suggestion
A serial recovery protocol is available over either a hardware serial port or a USB CDC ACM virtual serial port.
```

The SMP server implementation can be enabled by ``CONFIG_MCUBOOT_SERIAL=y``.
To set a type of an interface, use ``BOOT_SERIAL_DEVICE`` Kconfig choice, and select either ``CONFIG_BOOT_SERIAL_UART`` or ``CONFIG_BOOT_SERIAL_CDC_ACM`` value.
Which interface belongs to the protocol shall be set by the devicetree-chosen node:
- `zephyr,console` - If a hardware serial port is used.
- `zephyr,cdc-acm-uart` - If a virtual serial port is used.

### Entering the serial recovery mode

To enter the serial recovery mode, the device has to initiate rebooting, and a triggering event has to occur (for example, pressing a button).

By default, the serial recovery GPIO pin active state enters the serial recovery mode.

> Use the ``mcuboot_button0`` devicetree button alias to assign the GPIO pin to the MCUboot.

Reviewer: Note that all of this will become outdated once https://github.com/mcu-tools/mcuboot/pull/1629 is merged

Alternatively, MCUboot can wait for a limited time to check if DFU is invoked by receiving an MCUmgr command.
Select ``CONFIG_BOOT_SERIAL_WAIT_FOR_DFU=y`` to use this mode. ``CONFIG_BOOT_SERIAL_WAIT_FOR_DFU_TIMEOUT`` option defines

> the amount of time the device will wait for the trigger.

Reviewer: In seconds? ms?
